Narrow
%----------------------------------------------------------------------------------------
%	CHAP Seurat
%----------------------------------------------------------------------------------------

\chapterimage{blue-chapter-head_4-reduced.pdf} % Chapter heading image

\chapter{Seurat}\label{chap:Seurat}

\section{The Seurat language}

Since release 2.3.0, MetaR offers support for Single Cell RNA-Seq analysis with the Seurat 2.0 R package~\cite{butler2017integrated}. This functionality is implemented in the \textit{org.campagnelab.metar.seurat} language .

The \textit{org.campagnelab.metar.seurat} (Seurat) language consists of statements that help with the analysis of
single cell RNA sequencing data (scRNA-seq data). These statements cover a wide range of
functionality: loading scRNA-seq data, quality control/cleanup steps, adjusting data (by normalization
or scaling), plotting it, computing extra information based on the data (principal components, markers,
etc.), aligning data from multiple samples, performing limma analysis on it and other
functionality.

Seurat statements can be typed in an Analysis script (see Chapter~\ref{chap:Analyses}).
The simplest way to see what Seurat statements are valid at a given context in an
analysis, is to look at the suggestions offered by the context assistant. Figure~\ref{fig:ContextAssistantBeg}
and Figure~\ref{fig:ContextAssistantMid} show two examples of context assistants.
To activate the context assistant, just press
\keys{\return} in the script to generate an empty line. Placing the cursor on the empty line should
bring up the context assistant. 

\begin{remark}
If you do not see the context assistant, try pressing the space bar with the cursor positioned on an empty line.
Moreover, to see all possible statements, you can use auto-completion like for all other
statements in an analysis (see Chapter~\ref{chap:Analyses}).
\end{remark}


\begin{SCfigure}
  \centering
  \includegraphics[width=\figWidthSmall]{figures/ContextAssistantBeg.png}
    \caption[Context assistant at beginning of script.]{\textbf{Context assistant at
    beginning of script.} At the beginning of the script, the
    context assistant suggests loading a Seurat object either directly from the output
    of the 10X, or from an expression table.}
\label{fig:ContextAssistantBeg}
\end{SCfigure}

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/ContextAssistantMid.png}
    \caption[Context assistant after loading one Seurat object.]{\textbf{Context
    assistant after loading one Seurat object.} Once we have Seurat objects
    available in the script, more Seurat statements become valid.}
\label{fig:ContextAssistantMid}
\end{figure}

To start writing analyses using Seurat, you need to import devkit
\texttt{org\allowbreak.campagne\allowbreak{}lab\allowbreak.metaR}
and language \texttt{org\allowbreak.campagne\allowbreak{}lab\allowbreak.metar\allowbreak.seurat}.
\noindent The following sections describe the kinds of statements offered by the MetaR
\texttt{org\allowbreak.campagne\allowbreak{}lab\allowbreak.metar\allowbreak.seurat} language.

\section{The Seurat object}
A Seurat object is a structure that stores scRNA-seq data and associated information. Many of
the Seurat statements result in the creation of a new Seurat object. Seurat objects and
references to these objects are represented with a purple foreground. Moreover, you can
study the associated information of a Seurat object in the inspector window, when clicking
on any such object or a reference to it. Figure~\ref{fig:PropertiesSeurat} shows the
information (properties) associated with a newly created Seurat object. These properties
are used by some statements to assess whether the input Seurat object is ready for the computations
required by the statement.

\begin{SCfigure}
  \centering
  \includegraphics[width=\figWidthSmall]{figures/PropertiesSeurat.png}
    \caption[Properties view of a newly created Seurat object.]
    {\textbf{Properties view of a newly created Seurat object.} Many of the properties
    of this Seurat object are not yet computed. These properties are created automatically
    at the creation of the object.}
\label{fig:PropertiesSeurat}
\end{SCfigure}

\section{Loading Seurat objects}
There are two statements in the Seurat language for loading a Seurat object, one directly
from the files produced by 10X, and one from an expression matrix.

\subsection{Load 10X dataset}\label{subsec:Load10XDataset}
The \texttt{load 10X dataset} statement makes it possible to load a Seurat object directly
from the output files of 10X Genomics. The Seurat object and its properties become available
to the statements that follow the loading (until the line where it is deleted; see
Section~\ref{sec:OtherSeuratStatements}). You can create this statement by clicking
\menu{Load > Load scRNA-seq data from 10X output} in the context assistant, or by typing
the alias \texttt{load 10X dataset} on an empty line of Analysis (see Figure~\ref{fig:Load10XDataset})
for a new load 10X dataset statement).

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/Load10XDataset.png}
    \caption[New load 10X dataset statement.]{\textbf{New load 10X dataset statement.}
    Use this statement to load a Seurat object from the output of 10X Genomics.}
\label{fig:Load10XDataset}
\end{figure}

When creating a new \texttt{load 10X dataset} statement, the \texttt{by rejecting cell when}
strategy and \texttt{by rejecting gene when} come prefilled by default with certain upper
thresholds, but these strategies can be changed.

\paragraph{input directory} This field of \texttt{load 10X dataset} should point to
the directory produced by 10X Genomics cell ranger count output. Cellranger produces a directory structure which contains the following three files: ``barcodes.tsv'',
``genes.tsv'' and ``matrix.mtx''. Unless the directory you specify exists on the disk and it contains
these three files, the input directory field will be shown on a red background. Notice that
this field has a button to let you select the directory.

\paragraph{annotations} The annotations are references to group usages (see
Section~\ref{sec:ColumnGroupUsage}). They allow you to attach more information to
the data loaded from 10X Genomics. For instance, if you load data that corresponds
to a certain patient and a certain state of the sample (condition),
you would annotate the Seurat object with a reference to a group usage that represents the
patient and with a reference to a group usage that represents the state (see
Figure~\ref{fig:ExampleLoad10XDataset}).
These annotations are used by the expression tables generated from the Seurat object
in the limma statements (see Section~\ref{sec:LimmaSeurat}). Note that you
need to create groups and group usages in a Column Group Container root node
in the model of your solution. If a Column Group Container does not exist in
the model, see Section~\ref{sec:ColumnGroupContainer} to learn how to create one.

\paragraph{cleanup strategies}
There are three possible strategies available when loading data from 10X Genomics: rejecting
a cell whose number of genes is lower or higher than a threshold, rejecting a
gene when the number of cells it appears in is greater or smaller than a threshold, and
normalizing the data at loading time. The cleanup strategies are explained in
Section~\ref{sec:CleanupSeurat}.

\paragraph{output} The output of the \texttt{load 10X dataset} statement is a Seurat object.
You need to set the name of this object.

\paragraph{Example} Figure~\ref{fig:ExampleLoad10XDataset} presents an example of a
\texttt{load 10X dataset} statement.

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/ExampleLoad10XDataset.png}
    \caption[Example of load 10X dataset statement.]{\textbf{Example of load 10X dataset statement.}
    In this example, we load data for a patient denoted ``Patient1", from a sample of collapsed
    tubules tissue. As a result, we annotate this Seurat object with ``Patient1" and
    ``Collapsed". Furthermore, we reject the cells with less than 1000 genes expressed in
    them, and the genes that can only be found in less than 3 cells. Finally, we normalize
    the data using a scale factor of 10,000.}
\label{fig:ExampleLoad10XDataset}
\end{figure}

\subsection{Load dataset from table}
The \texttt{load dataset from} statement makes it possible to load a Seurat object from a
table that contains expression data. You can create this statement by clicking
\menu{Load > Load scRNA-seq data from table} in the context assistant, or by typing
the alias \texttt{load dataset from table} on an empty line of Analysis. This statement
is identical in fields and behavior to the \texttt{load 10X dataset} statement, except
for the place of input (one is a directory generated by 10X Genomics, and one is a table).
Thus, we refer you to Subsection~\ref{subsec:Load10XDataset} for further details on the common fields.
% TODO: describe input table format.

\paragraph{input table}
The input table should have genes on the rows and cells on the columns.
Moreover, the table should have the row names. To reference a table
from this statement, you need to have a table available in the model (either by importing
it, or by obtaining it from other statements; see Section~\ref{chap:Tables}).

\paragraph{Example} Figure~\ref{fig:ExampleLoadTable} presents an example of a
\texttt{load data from table} statement.

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/ExampleLoadTable.png}
    \caption[Example of load data from table statement.]{\textbf{Example of load data from table statement.}
    In this example, we load data from a table with simulated data for a patient
    denoted ``Patient1", and from a sample of collapsed tubules tissue.
    As a result, we annotate the Seurat object with ``Patient1" and
    ``Collapsed". Furthermore, we reject the cells with less than 1500 genes expressed in
    them, and the genes that can only be found in less than 4 cells. We do not normalize
    the data at loading time.} 
\label{fig:ExampleLoadTable}
\end{figure}

\section{QC and Clean Up}\label{sec:CleanupSeurat}
One important step in the analysis of scRNA-seq data is quality control on the data. In
Seurat, you can do that with the \texttt{cleanup seurat} statement. Usually, this
quality control step is done after observing some diagnostic plots on the data 
(see Section~\ref{subsec:DiagnosticPlots} for diagnostic plots).

You can create this statement by clicking \menu{Cleanup} in the context assistant,
or by typing the alias \texttt{cleanup seurat} on an empty line of Analysis (see
Figure~\ref{fig:CleanupSeurat} for a new \texttt{cleanup seurat} statement). Note that in
the context assistant, you always need to specify a strategy that is going to be instantiated
in the \texttt{cleanup seurat} statement.

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/CleanupSeurat.pdf}
    \caption[New cleanup seurat statement.]{\textbf{New cleanup seurat statement.}
    Use this statement to perform quality control on a Seurat object.}
\label{fig:CleanupSeurat}
\end{figure}

\paragraph{input Seurat} You need to specify the Seurat object on which quality control
is done. Only Seurat objects that are defined in previous statements are visible in the
current cleanup statement, due to scoping rules.

\paragraph{output Seurat} The statement creates a new Seurat object that represents the
input Seurat object with the modifications prescribed in the cleanup statement. The
\texttt{cleanup seurat} statement creates a default name for the output Seurat object,
but this name can be modified.

\paragraph{strategies} To introduce a strategy, you have to press \keys{\ctrl+\space}
in the red space on the second line of the statement. All the available strategies will
subsequently appear in the menu. In the next sections, we explain these strategies. Two
of these strategies are only available in the initial cleanup section of the loading
statements: the reject gene strategy and the normalization strategy.

\subsection{Reject gene strategy}
The \texttt{by rejecting gene when} strategy specifies a comparison operation where the
\texttt{number of cells where gene is expressed in} is compared to a threshold.
This strategy filters the input Seurat object by dropping the genes for which
the comparison is true. For this strategy, the left-hand side of the comparison can be
only \texttt{number of cells where gene is expressed in}.

\subsection{Reject cell strategy}
Another strategy is when you reject cells based on whether certain conditions are fulfilled.
Pressing \keys{\ctrl+\space} on the left-hand side of a condition in this strategy gives you
three possibilities: number of UMIs (unique molecular identifiers), number of genes and
percentage of mitochondrial genes in cell. This strategy filters the input Seurat object
by dropping the cells for which any of the comparisons is true. You can specify multiple
conditions in this strategy (see Figure~\ref{fig:ExampleCleanupSeurat} for an example). Note
that the threshold for mitochondrial genes is an integer number representing the percentage. 

\subsection{Regress out strategy}
Because single cell datasets often contain technical noise, batch effects, or even undesired
biological sources of variation, the cleanup statement offers a strategy to regress out
these sources of variation. You can write a list of such sources to be regressed out;
after introducing one element inside the square brackets of the regress out strategy, just
press \keys{\return} to introduce a new element (see Figure~\ref{fig:ExampleCleanupSeurat}
for an example). Note that this strategy also scales the data after regressing out the
variation, so it is adviced that it is the last strategy in the cleanup statement. As a result,
we obtain scaled data after the cleanup.

\subsection{Accept highly variable genes strategy}
It is common that you focus on the highly variable genes for downstream analysis; for this end,
you can use the \texttt{by accepting highly variable genes when} strategy. This will mark
the highly variable genes in the Seurat object. The highly variable genes are used by further Seurat
statements. To compute the highly variable genes, Seurat makes use of the average expression
and dispersion of each gene. The strategy allows you to set the cutoff for the average expression
and dispersion axis. The cutoff can be typed inside the square brackets of the strategy. To
introduce a new cutoff, just type \keys{\return} after the last one introduced, and press
\keys{\ctrl+\space} to see the available choices (see Figure~\ref{fig:ExampleCleanupSeurat}
for an example). Note that once you specify this strategy in the cleanup statement, an
output plot is pops up, that represents the dispersion versus average expression plot. You
need to provide a name for this plot.

\subsection{Normalization strategy}
Another strategy that can be used only in the cleanup section of the loading statements is the
normalization strategy. This strategy normalizes the data using the scale factor provided
in the strategy and log-transforms the result. There is also a separate statement for normalization (see
Subsection~\ref{sec:NormalizeSeurat}).

\paragraph{Example} Figure~\ref{fig:ExampleCleanupSeurat} presents an example of a
\texttt{cleanup seurat} statement.

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/ExampleCleanupSeurat.png}
    \caption[Example of cleanup seurat statement.]{\textbf{Example of cleanup seurat statement.}
    This illustrates almost all the features of the cleanup statement. We set a cutoff for
    all the axes, we reject cells based on three different conditions and we regress out
    unwanted variations for two different variables.}
\label{fig:ExampleCleanupSeurat}
\end{figure}

\section{Adjusting Seurat objects}
There are two statements in Seurat that are meant for standard adjusting
(pre-processing) the data: normalization and scaling. Some statements require that the
Seurat object is already normalized or scaled. For instance, the calculation of highly
variable genes needs a normalized Seurat object as input.

\subsection{Normalize Seurat object}\label{sec:NormalizeSeurat}
The normalization statement normalizes the expression data stored in the Seurat object and
log-transforms the result. You can create this statement by clicking
\menu{Adjust > Normalize} in the context assistant, or by typing
the alias \texttt{normalize seurat} on an empty line of Analysis (see Figure~\ref{fig:NormalizeSeurat}
for an example of a new \texttt{normalize seurat} statement).

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/NormalizeSeurat.pdf}
    \caption[New normalize seurat statement.]{\textbf{New normalize seurat statement.}
    A new \texttt{normalize seurat} statement with three fields to be completed.}
\label{fig:NormalizeSeurat}
\end{figure}

\paragraph{input Seurat}
The input Seurat object can be any Seurat object created in a previous statement, but it
should not be an already normalized Seurat object. You will receive an error if that is the
case.

\paragraph{scale factor}
You also need to specify the scaling factor for the normalization.

\paragraph{output Seurat}
The output is a normalized Seurat object. You need to introduce a name for
it.

\subsection{Scale Seurat object}
The scaling statement scales the data stored in the Seurat object.
You can create this statement by clicking \menu{Adjust > Scale} in the context assistant,
or by typing the alias \texttt{scale seurat} on an empty line of Analysis (see Figure~\ref{fig:ScaleSeurat}
for an example of a new \texttt{scale seurat} statement).

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/ScaleSeurat.pdf}
    \caption[New scale seurat statement.]{\textbf{New scale seurat statement.}
    A new \texttt{scale seurat} statement with two fields to be completed.}
\label{fig:ScaleSeurat}
\end{figure}

\paragraph{input Seurat}
The input Seurat object can be any Seurat object created in a previous statement. Press
\keys{\ctrl+\space} to see the available choices.

\paragraph{output Seurat}
The output is a scaled Seurat object. You need to introduce a name for it.

\section{Plotting Seurat objects}
While there are statements in Seurat that create plots as an auxiliary artifact,
there are also dedicated statements to create plots. These plots are
usually used to interactively explore the data and decide on parameters for the next
statements in the analysis.

\subsection{Diagnostic plots}\label{subsec:DiagnosticPlots}
One important statement, that is usually used directly after loading a Seurat object, is the
\texttt{Diagnostic plots} statement. You can create this statement by clicking
\menu{Plots > Diagnostic plots} in the context assistant, or by typing
the alias \texttt{Diagnostic plots} on an empty line of Analysis (see Figure~\ref{fig:DiagnosticPlots}
for an example of a new \texttt{Diagnostic plots} statement).

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/DiagnosticPlots.pdf}
    \caption[New Diagnostic Plots statement.]{\textbf{New Diagnostic Plots statement}}
\label{fig:DiagnosticPlots}
\end{figure}

\paragraph{input Seurat} You need to specify the name of the input Seurat object. Press
\keys{\ctrl+\space} to see the available Seurat objects.

\paragraph{width and height} You can also specify the width and height of the five plots,
all at once, by modifying these two parameters.

\paragraph{output plots} There are five output plots from this statement. Three of them
are violin plots for the number of genes in the cell (nGene), number of unique molecular identifier
in the cell (nUMI) and the percentage of mitochondrial genes in the cell (percent.mito). The other two are
scatter plots showing percent.mito versus nUMI, and nGene versus nUMI. These five plots
have default names, but you can change these names by clicking on the name and typing. To
visualize these plots, you can use the \texttt{multiplot} statement (see Subsection~\ref{subsec:Multiplot}).

These five plots are typically used to decide on thresholds for the rejection
of cells or other parameters of the cleanup statement.

\subsection{Features plot}
The features plot highlights the specified features on a plot with tSNE clusters. This statement
is useful in exploring markers for the clusters.
You can create this statement by clicking
\menu{Plots > Features plot} in the context assistant, or by typing
the alias \texttt{Feature plot} on an empty line of Analysis (see Figure~\ref{fig:FeaturesPlot}
for an example of a new \texttt{Feature plot} statement).

\begin{figure}[h!tbp]
  \centering
  \includegraphics[width=\figWidthWide]{figures/FeaturesPlot.png}
    \caption[New feature plot statement.]{\textbf{New feature plot statement.} A new
    \texttt{Feature plot} statement expects a list of features, an input Seurat object
    and a name for the output Seurat object.}
\label{fig:FeaturesPlot}
\end{figure}

\paragraph{input Seurat} The input Seurat object that has the tSNE computed already. If this
is not the case, an error will be reported.

\paragraph{features} You need to specify a list of features that will be highlighted in the
clusters. The output plot is actually a collection of smaller plots, one per feature; each
smaller plot highlights one feature in the clusters. To enter feature names, press \keys{\return}
inside the square brackets of the statement, and type in a name. To introduce another
feature in the list, press \keys{\return} at the end of a feature.

\paragraph{output plot} You need to introduce the name of the output plot.

\subsection{Features and total plot}
The features and total plot highlights the specified features and their cumulated effect
(obtained by multiplying the expressions of these genes) on a plot with tSNE clusters. This statement
is useful to see whether the specified features characterize a cluster together.
You can create this statement by clicking
\menu{Plots > Features and total plot} in the context assistant, or by typing
the alias \texttt{Feature plot and total} on an empty line of Analysis (see Figure~\ref{fig:FeaturesPlotTotal}
for an example of a new \texttt{Feature plot and total} statement). The output plot
for such a statement where features ``PLEK" and ``SDPR" are given as input, can be seen in
Figure~\ref{fig:TwoFeaturesPlot}.

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/FeaturesPlotTotal.png}
    \caption[New feature plot and total statement.]{\textbf{New feature plot and total statement.} A new
    \texttt{Feature plot and total} statement expects a list of features, an input Seurat object
    and a name for the output Seurat object.}
\label{fig:FeaturesPlotTotal}
\end{figure}

\begin{SCfigure}
  \centering
    \includegraphics[width=\figWidthNarrow]{figures/TwoFeaturesPlot.pdf}
    \caption[Two features and their cumulated expression highlighted.]{\textbf{Two features
    and their cumulated expression highlighted.} This figure shows plots for two features (genes),
    and their cumulated expression (Total).}
\label{fig:TwoFeaturesPlot}
\end{SCfigure}

The parameters for this statement are identical to the parameters for the \texttt{Feature plot}
statement.

\section{Adding information to Seurat objects}
Another important step in the analysis of scRNA-seq data is the computation of principal
components, of tSNE clusters and of the markers per cluster. There are three dedicated
statements in Seurat, for the three functionalities.

\subsection{Add principal component information}
The \texttt{add principal components} statement computes the principal components for the expression
data in the input Seurat object. You can create this statement by clicking
\menu{Add > Principal component info} in the context assistant, or by typing
the alias \texttt{add principal components} on an empty line of Analysis (see Figure~\ref{fig:AddPCA}
for an example of a new \texttt{add principal component} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/AddPCA.pdf}
    \caption[New add principal components statement.]{\textbf{New add principal components statement.} A new
    \texttt{add principal components} statement expects an input Seurat object, a name for the output Seurat object
    and a name for the auxiliary plot produced.}
\label{fig:AddPCA}
\end{figure}

\paragraph{input Seurat} The input Seurat object for which principal component analysis is added.

\paragraph{output Seurat} The output Seurat object will have the principal component analysis
property set to true. You can change the name of this object.

\paragraph{plot} This statement produces a plot as an auxiliary artifact. The plot shows the
standard deviations of the principal components. This plot helps you decide which PCs to use
further in the analysis, by choosing the cutoff where there is an elbow in the plot. For
an example, see Figure~\ref{fig:ElbowView}.

\begin{SCfigure}
  \centering
    \includegraphics[width=\figWidthNarrow]{figures/ElbowView.pdf}
    \caption[The standard deviations of principal components plot.]{\textbf{The standard deviations of
    principal components plot.} Given this figure, we would choose 17 as the cutoff value,
    because it resides at the elbow of the plot.}
\label{fig:ElbowView}
\end{SCfigure}

\subsection{Add clusters information}
The \texttt{add clusters} statement computes the tSNE clusters for the expression
data in the input Seurat object. This statement needs PC information, so it needs to be
run after the \texttt{add principal components} statement. You can create this statement by clicking
\menu{Add > Clusters info} in the context assistant, or by typing
the alias \texttt{add clusters} on an empty line of Analysis (see Figure~\ref{fig:AddClusters}
for an example of a new \texttt{add clusters} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/AddClusters.png}
    \caption[New add clusters statement.]{\textbf{New add clusters statement.} A new
    \texttt{add clusters} statement produces an auxiliary plot showing the tSNE clusters
    and uses information from the principal components analysis via the parameters captured
    in range of PC.}
\label{fig:AddClusters}
\end{figure}

\paragraph{input Seurat} The input Seurat object for which tSNE clusters information is added.

\paragraph{range of PCs} Based on the cutoff you see in the plot produced by the principal
component analysis statement, you fill in the range of PCs to use in the tSNE analysis.

\paragraph{resolution} This parameter affects the number of clusters the tSNE analysis will
produce (the larger the resolution, the more clusters it produces).
A typical value for this parameter is 0.6.

\paragraph{output Seurat} The output Seurat object will have the tSNE property set to true.

\paragraph{plot} This statement also produces a plot that shows the tSNE clusters.

\subsection{Add markers information}
Once clusters of cells have been identified, you might want to find genes whose expression is specific of the cluster. Seurat calls these genes cluster markers. The \texttt{add markers} statement computes the markers for each cluster, so it needs to be
run on an input Seurat object that has tSNE clustering information.
You can create this statement by clicking
\menu{Add > Markers per cluster info} in the context assistant, or by typing
the alias \texttt{add markers} on an empty line of Analysis (see Figure~\ref{fig:AddMarkers}
for an example of a new \texttt{add markers} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/AddMarkers.png}
    \caption[New add markers statement.]{\textbf{New add markers statement.} A new
    \texttt{add markers} statement produces an auxiliary table with the identified markers, for each cluster.}
\label{fig:AddMarkers}
\end{figure}

\paragraph{input Seurat} The input Seurat object for which markers information is added.

\paragraph{output Seurat} The output Seurat object that will have the markers information.

\paragraph{number of markers} This parameters allows you to specify how many markers per cluster
you want to compute.

\paragraph{xFold} This parameter specifies the minimum log-fold difference between
the two compared groups of cells when computing the markers.

\paragraph{percentage} This parameter specifies what is the minimum percentage of cells
in either of the two groups of cells, where genes need to be detected when computing
the markers.

\paragraph{table} This statement also produces a table that contains the markers per cluster.
You can see the columns of the table by clicking on it and looking in the \inspectorTabIcon tab.

\section{Aligning Seurat objects}
One common scenario in the scRNA-seq data analysis, is the comparison of different
samples. Before doing comparisons between these different samples, it is important to align them.
To this end, Seurat 2.0+ provides an approach to align cells found in two Seurat objects. MetaR wraps this functionality in the following analysis statements. 
\subsection{Prealign Seurat objects}
The pre-align statement prepares the Seurat objects for alignment and it outputs a plot that
is used to decide on parameters for the alignment statement. You can create this statement by clicking
\menu{Align > Prealign} in the context assistant, or by typing
the alias \texttt{prealign seurats} on an empty line of Analysis (see Figure~\ref{fig:PrealignSeurat}
for an example of a new \texttt{prealign seurats} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/PrealignSeurat.pdf}
    \caption[New prealign statement.]{\textbf{New prealign statement.} The \texttt{prealign
    seurats} statement outputs a plot showing the state of the first two canonical correlation vectors
    (used under the hood), and also heatmaps for as many of these vectors as you specify in the input
    parameter.}
\label{fig:PrealignSeurat}
\end{figure}

\paragraph{input seurats} This statement expects two input Seurat objects that will be aligned
by this statement.

\paragraph{heatmaps for canonical correlation vectors up to} The canonical correlation (CC) vectors specified in the
range of this statement are shown in heatmaps so that you can assess what CCs to use further
in the analysis, in particular, for the alignment. You can decide to choose up to the latest CC
that shows some variation. Figure~\ref{fig:HeatmapxsCCs} shows an example.

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthNarrow]{figures/HeatmapCCs.pdf}
    \caption[Heatmaps showing the CC vectors and their scores per gene.]{\textbf{Heatmaps showing
    the CC vectors and their scores per gene.} For this particular example, we choose CCs
    up to 13, because this is where the signal starts to fade.}
\label{fig:HeatmapxsCCs}
\end{figure}

\paragraph{output Seurat} The Seurat object that is going to contain the aligned result.

\paragraph{CCA plot} This plot shows information from the first two CC vectors so that
you can assess whether the alignment can be made.

\paragraph{heatmaps plot} This plot  contains the heatmaps for the CC vectors specified at
the input. These heatmaps can be used to decide what CCs to use further down in the analysis.

\subsection{Align Seurat object}
The \texttt{align seurat} statement aligns the two Seurat objects provided in the pre-alignment.
You can create this statement by clicking
\menu{Align > Align} in the context assistant, or by typing
the alias \texttt{align seurat} on an empty line of Analysis (see Figure~\ref{fig:AlignSeurat}
for an example of a new \texttt{align seurat} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/AlignSeurat.pdf}
    \caption[New align statement.]{\textbf{New align statement.} A new align statement
    needs information obtained from the prealign statement and outputs two auxiliary
    tSNE plots.}
\label{fig:AlignSeurat}
\end{figure}

\paragraph{input Seurat} The input Seurat should be the Seurat object obtained from the
\texttt{prealign seurats} statement.

\paragraph{CCA dimensions from - to} In these parameters, you specify the CCs that you have
chosen as a result of inspecting the heatmaps from the \texttt{prealign seurats} statement.

\paragraph{output Seurat} The output Seurat contains aligned data from the two Seurat objects
given as input to the prealign statement.

\paragraph{tSNE plots} This statement generates two plots with the tSNE clusters, one where
the two datasets from the two input Seurat objects are highlighted, and one where the clusters
themselves are highlighted.

\section{Limma for Seurat objects}\label{sec:LimmaSeurat}
In Seurat, there are special statements used for the differential expression analysis
using the limma package.

\subsection{Pre-limma Seurat object}
The pre-limma statement extracts an aggregated count table from the Seurat object given as input. This is needed to obtain counts for genes in individual clusters. The procedure simply sums the counts of a gene across cells of a cluster, while preserving experimental conditions. To this end, we aggregate columns per individual Seurat object (we kept track of Seurat objects that were merged, or aligned)and per tSNE cluster. You can create this statement by clicking \menu{Limma > Prelimma} in the context assistant, or by typing the alias \texttt{pre limma} on an empty line of Analysis (see Figure~\ref{fig:PreLimmaSeurat} for an example of a new \texttt{pre limma} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/PreLimmaSeurat.pdf}
    \caption[New pre limma statement.]{\textbf{New pre limma statement.} The \texttt{pre
    limma} statement generates an aggregated count table and needs tSNE cluster information
    as input.}
\label{fig:PreLimmaSeurat}
\end{figure}

\paragraph{input Seurat} The input is a Seurat object for which we want to compute the
differentially expressed genes.

\paragraph{tSNE clusters from - to} These parameters need to be filled in with the clusters
that result from the tSNE computation.

\paragraph{output table} The aggregated count table contains one column per input Seurat
object and per cluster. For instance, if we merged $\mathit{P1C}$ and
$\mathit{P1D}$ into $\mathit{P1}$, $\mathit{P2C}$ and $\mathit{P2D}$ into $\mathit{P2}$, we aligned
$\mathit{P1}$ and $\mathit{P2}$ into $\mathit{P1P2}$, and we have obtained clusters
$\mathit{0}$ and $\mathit{1}$, then we get columns $\mathit{P1C0}$, $\mathit{P1C1}$, $\mathit{P1D0}$, etc.
Besides these columns, the output table contains a column with gene names and has also row names.

The annotations that were introduced when loading the Seurat objects will be propagated to the output table.

\subsection{Limma voom}
The limma statement computes the differentially expressed genes for the input count table and outputs
a table with results per contrast. You can create this statement by clicking
\menu{Limma > Limma} in the context assistant, or by typing
the alias \texttt{limma voom} on an empty line of Analysis (see Figure~\ref{fig:LimmaSeurat}
for an example of a new \texttt{limma voom} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/LimmaSeurat.pdf}
    \caption[New limma statement.]{\textbf{New limma statement.} A limma statement applied on
    a table containing expression data.}
\label{fig:LimmaSeurat}
\end{figure}

\paragraph{input table} The input table should be a table containing expression data. Moreover,
the table needs to be annotated with some group usages that will subsequently be used in the
model formula and in contrasts.

\paragraph{output tables} For each contrast specified in this statement, there will be one
output results table containing information such as the log-fold change and adjusted P-value.
Moreover, each of the output tables comes with a path to a directory where an interactive
web plot is generated.

For the explanation of the other parameters of the limma statement, look at the parameters described in the
main limma statement provided by MetaR in Chapter~\ref{chap:LimmaVoom}. The only difference is that
in the Seurat statement, one can specify multiple contrasts (under the \texttt{comparisons} attribute).

\section{Other Seurat statements}\label{sec:OtherSeuratStatements}
There are two more Seurat statements that are useful at various places in the analysis: the
merge and the delete statements.

\subsection{Merge Seurat objects}
The merge statement takes two Seurat objects as input and creates one single output Seurat
object as output. Under the hood, this statement merges the genes in the two Seurat objects
and keeps all the cells from the two objects.
You can create this statement by clicking \menu{Merge} in the context assistant, or by typing
the alias \texttt{merge seurat object} on an empty line of Analysis (see Figure~\ref{fig:MergeSeurat}
for an example of a new \texttt{merge seurat objects} statement).

\begin{figure}[h!tbp]
  \centering
    \includegraphics[width=\figWidthWide]{figures/MergeSeurat.pdf}
    \caption[New merge seurat objects statement.]{\textbf{New merge seurat objects statement.} A merge statement
    needs two input Seurat objects and returns them merged in the output Seurat object.}
\label{fig:MergeSeurat}
\end{figure}

\paragraph{input Seurat objects} You need to provide two input Seurat objects.

\paragraph{output Seurat object} The output Seurat objects contains all the information
from both input Seurat objects.

\subsection{Delete Seurat object}
This statement deletes the provided Seurat object. After this statement, the input Seurat
object is not available as input anymore. This statement was introduced to save memory.
You can create this statement by clicking
\menu{Delete} in the context assistant, or by typing
the alias \texttt{delete seurat} on an empty line of Analysis (see Figure~\ref{fig:DeleteSeurat}
for an example of a new \texttt{delete} statement).

\begin{SCfigure}
  \centering
    \includegraphics[width=\figWidthTiny]{figures/DeleteSeurat.pdf}
    \caption[New delete statement.]{\textbf{New delete statement.} A delete statement needs only the
    Seurat object to be deleted as input.}
\label{fig:DeleteSeurat}
\end{SCfigure}

\paragraph{input Seurat} You need to introduce the Seurat object that needs to be deleted.
